Examples/c++/Linked List/DLList.h

```diff
@@ -0,0 +1,401 @@
+		NodePtr Next = nullptr;
+		NodePtr Previous = nullptr;
+
+		DLNode() = default;
+		DLNode(const T& data, NodePtr prev, NodePtr next)
+			: Data(data), Previous(prev), Next(next) {}
+		DLNode(T&& data, NodePtr prev, NodePtr next)
+			: Data(std::move(data)), Previous(prev), Next(next) {}
+	};
+
+	template<typename T>
+	class DLList
+	{
+	public:
+		using ValueType = T;
+		using Node = DLNode<T>;
+		using NodePtr = DLNode<T>*;
+		using Iterator = DLLIterator<DLList<T>>;
+	public:
+		DLList()
+		{
+			m_SentinelNode = new Node();
+			m_SentinelNode->Next = m_SentinelNode;
+			m_SentinelNode->Previous = m_SentinelNode;
+		}
+
+		~DLList()
+		{
+			Clear();
+			delete m_SentinelNode;
+		}
+
+		void PushFront(const T& data)
+		{
+			InsertAfter(m_SentinelNode, data);
+		}
+
+		void PushFront(T&& data)
+		{
+			InsertAfter(m_SentinelNode, std::move(data));
+		}
+
+		template<typename... Args>
+		T& EmplaceFront(Args&&... args)
+		{
+			return EmplaceAfter(m_SentinelNode, std::forward<Args>(args)...);
+		}
+
+		void PushBack(const T& data)
+		{
+			InsertBefore(m_SentinelNode, data);
+		}
+
+		void PushBack(T&& data)
+		{
+			InsertBefore(m_SentinelNode, std::move(data));
+		}
+
+		template<typename... Args>
+		T& EmplaceBack(Args&&... args)
+		{
+			return EmplaceBefore(m_SentinelNode, std::forward<Args>(args)...);
+		}
+
+		void Insert(Iterator where, const T& data)
+		{
+			InsertBefore((NodePtr)where, data);
+		}
+
+		void Insert(Iterator where, T&& data)
+		{
+			InsertBefore((NodePtr)where, std::move(data));
+		}
+
+		template<typename... Args>
+		T& Emplace(Iterator where, Args&&... args)
+		{
+			return EmplaceBefore((NodePtr)where, std::forward<Args>(args)...);
+		}
+
+		void PopFront()
+		{
+			DeleteNode(m_SentinelNode->Next);
+		}
+
+		void PopBack()
+		{
+			ASSERT(!Empty(), "PopBack() called on an empty list!");
+			DeleteNode(m_SentinelNode->Previous);
+		}
+
+		void Erase(Iterator where)
+		{
+			ASSERT((NodePtr)where != m_SentinelNode, "Invalid Iterator");
+			DeleteNode(where);
+		}
+
+		void MoveToBack(Iterator from, Iterator to)
+		{
+			if (m_Size <= 1 || to == end())
+				return;
+
+			NodePtr startNode = (NodePtr)from;
+			NodePtr endNode = ((NodePtr)to)->Previous;
+
+			NodePtr startNodePrevious = startNode->Previous;
+
+			endNode->Next = m_SentinelNode;
+			m_SentinelNode->Previous->Next = startNode;
+			startNode->Previous = m_SentinelNode->Previous;
+
+			m_SentinelNode->Previous = endNode;
+			startNodePrevious->Next = (NodePtr)to;
+			((NodePtr)to)->Previous = startNodePrevious;
+
+		}
+
+		void Reverse()
+		{
+			NodePtr current = m_SentinelNode->Next;
+
+			while (current != m_SentinelNode)
+			{
+				std::swap(current->Next, current->Previous);
+				current = current->Previous;
+			}
+
+			std::swap(m_SentinelNode->Next, m_SentinelNode->Previous);
+			
+			m_SentinelNode->Previous->Next = m_SentinelNode;
+			m_SentinelNode->Next->Previous = m_SentinelNode;
+		}
+
+		void Reverse(Iterator from, Iterator to)
```

Comment:
Why having several functions which are doing the same without reusing each other. It's making the code difficult to maintain as if a bug is found at one of the function the other should also be checked

Reply:
The Reverse() function override with no parameters is not necessarily needed so it will be removed.